File: c-pq/src/fib_heap_pq.c
#include "fib_heap_pq.h"
#include "constants.h"
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <math.h>

void _merge_with_root_list(FibHeapPQ *pq, FibHeapNode *node) {
    if (pq->min == NULL) {
        node->left = node;
        node->right = node;
        pq->min = node;
    } else {
        node->right = pq->min;
        node->left = pq->min->left;

        pq->min->left->right = node;
        pq->min->left = node;

        node->right->left = node;
    }

    if (node->priority < pq->min->priority) {
        pq->min = node;
    }
}

void _remove_from_child_list(FibHeapNode *parent, FibHeapNode *node) {
    if (parent->child == node) {
        parent->child = (node->right == node) ? NULL : node->right;
    }
    node->left->right = node->right;
    node->right->left = node->left;
    node->left = node;
    node->right = node;
    parent->degree--;
}

void _remove_from_root_list(FibHeapPQ *pq, FibHeapNode *node) {
    if (node->right == node) {
        pq->min = NULL;
    } else {
        if (pq->min == node) {
            pq->min = node->right;
        }

        node->left->right = node->right;
        node->right->left = node->left;
    }

    node->left = node;
    node->right = node;
}

void _cut(FibHeapPQ *pq, FibHeapNode *l_node, FibHeapNode *r_node) {
    _remove_from_child_list(r_node, l_node);
    _merge_with_root_list(pq, l_node);
    l_node->parent = NULL;
    l_node->marked = false;
}

void _cascading_cut(FibHeapPQ *pq, FibHeapNode *node) {
    FibHeapNode *ref_node = node->parent;
    if (ref_node != NULL) {
        if (!node->marked) {
            node->marked = true;
        } else {
            _cut(pq, node, ref_node);
            _cascading_cut(pq, ref_node);
        }
    }
}

void _heap_link(FibHeapPQ *pq, FibHeapNode *l_node, FibHeapNode *r_node) {
    _remove_from_root_list(pq, l_node);
    l_node->left = l_node;
    l_node->right = l_node;
    l_node->parent = r_node;
    l_node->marked = false;

    if (r_node->child == NULL) {
        r_node->child = l_node;
    } else {
        l_node->right = r_node->child;
        l_node->left = r_node->child->left;
        r_node->child->left->right = l_node;
        r_node->child->left = l_node;
    }
    r_node->degree++;
}

void _consolidate(FibHeapPQ *pq) {
    if (pq->min == NULL || pq->min->right == pq->min) {
        return;
    }

    size_t max_degree = log2(pq->size) * 2 + 1;
    FibHeapNode *A[max_degree];
    memset(A, 0, sizeof(A));

    size_t root_count = 0;
    FibHeapNode *w = pq->min;
    do { root_count++; w = w->right; } while (w != pq->min);

    FibHeapNode *roots[root_count];
    w = pq->min;
    for (size_t i = 0; i < root_count; i++) {
        roots[i] = w;
        w = w->right;
    }

    for (size_t i = 0; i < root_count; i++) {
        FibHeapNode *x = roots[i];
        size_t d = x->degree;
        while (A[d] != NULL) {
            FibHeapNode *y = A[d];
            if (x->priority > y->priority) {
                FibHeapNode *tmp = x; x = y; y = tmp;
            }
            _heap_link(pq, y, x);
            A[d] = NULL;
            d++;
        }
        A[d] = x;
    }

    pq->min = NULL;
    for (size_t i = 0; i < max_degree; i++) {
        if (A[i] == NULL) {
            continue;
        }
        A[i]->left = A[i]->right = A[i];
        if (pq->min == NULL) {
            pq->min = A[i];
        } else {
            A[i]->right = pq->min->right;
            A[i]->left = pq->min;
            pq->min->right->left = A[i];
            pq->min->right = A[i];
            if (A[i]->priority < pq->min->priority) {
                pq->min = A[i];
            }
        }
    }
}

static void _free_node_list(FibHeapNode *start) {
    if (start == NULL) return;
    FibHeapNode *cur = start;
    do {
        FibHeapNode *next = cur->right;
        _free_node_list(cur->child);
        free(cur);
        cur = next;
    } while (cur != start);
}

FibHeapPQ *fh_new(void) {
    FibHeapPQ *new_pq = malloc(sizeof(FibHeapPQ));
    if (new_pq == NULL) {
        printf("%s", ALLOCATION_ERROR);
        exit(1);
    }

    new_pq->min = NULL;
    new_pq->size = 0;

    return new_pq;
}

FibHeapNode *fh_insert(FibHeapPQ *pq, void *data, size_t priority) {
    FibHeapNode *new_node = malloc(sizeof(FibHeapNode));
    if (new_node == NULL) {
        printf("%s", ALLOCATION_ERROR);
        exit(1);
    }

    new_node->priority = priority;
    new_node->degree = 0;
    new_node->marked = false;

    new_node->parent = NULL;
    new_node->child = NULL;
    new_node->left = new_node;
    new_node->right = new_node;

    _merge_with_root_list(pq, new_node);

    new_node->data = data;

    pq->size++;

    return new_node;
}

void *fh_extractMin(FibHeapPQ *pq) {
    FibHeapNode *ref_node = pq->min;
    if (ref_node != NULL) {
        FibHeapNode *child = ref_node->child;
        if (child != NULL) {
            FibHeapNode *start = child;
            FibHeapNode *next;

            do {
                next = child->right;

                child->left = child;
                child->right = child;

                _merge_with_root_list(pq, child);

                child->parent = NULL;
                child->marked = false;
                child = next;
            } while (child != start);
        }

        bool singleton = (ref_node->right == ref_node);
        FibHeapNode *next_root = ref_node->right;

        _remove_from_root_list(pq, ref_node);

        if (singleton) {
            pq->min = NULL;
        } else {
            pq->min = next_root;
            _consolidate(pq);
        }

        pq->size--;
    } else {
        return NULL;
    }

    void *data = ref_node->data;
    free(ref_node);
    return data;
}

bool fh_isEmpty(FibHeapPQ *pq) { return pq->size == 0; }

void *fh_remove(FibHeapPQ *pq, FibHeapNode *node) {
    node->priority = 0;

    FibHeapNode *parent = node->parent;
    if (parent != NULL) {
        _cut(pq, node, parent);
        _cascading_cut(pq, parent);
    }
    pq->min = node;
    return fh_extractMin(pq);
}

void fh_decreaseKey(FibHeapPQ *pq, FibHeapNode *node, size_t new_priority) {
    if (new_priority > node->priority) {
        return;
    }

    node->priority = new_priority;
    FibHeapNode *ref_node = node->parent;

    if (ref_node != NULL && node->priority < ref_node->priority) {
        _cut(pq, node, ref_node);
        _cascading_cut(pq, ref_node);
    }
    if (node->priority < pq->min->priority) {
        pq->min = node;
    }
}

FibHeapNode *fh_peek(FibHeapPQ *pq) { return pq->min; }

FibHeapPQ *fh_merge(FibHeapPQ *lhs, FibHeapPQ *rhs) {
    FibHeapPQ *new_pq = malloc(sizeof(FibHeapPQ));
    if (new_pq == NULL) {
        printf("%s", ALLOCATION_ERROR);
        exit(1);
    }

    // HACK: the new_pq->min value will be set in the _merge_with_root_list function
    new_pq->min = lhs->min;
    new_pq->size = lhs->size + rhs->size;

    if (lhs->min == NULL || rhs->min == NULL) {
        new_pq->min = lhs->min ? rhs->min : lhs->min;
        return new_pq;
    }

    // Splice the two circular lists together directly
    FibHeapNode *lhs_last = lhs->min->left;
    FibHeapNode *rhs_last = rhs->min->left;

    lhs_last->right = rhs->min;
    rhs->min->left = lhs_last;
    rhs_last->right = new_pq->min;
    new_pq->min->left = rhs_last;

    if (rhs->min != NULL &&
        (new_pq->min == NULL || rhs->min->priority < new_pq->min->priority)) {
        new_pq->min = rhs->min;
    }

    return new_pq;
}

void fh_free(FibHeapPQ *pq) {
    _free_node_list(pq->min);
    free(pq);
}

File: rust-pq/src/fib_heap_pq.rs
use std::ptr::NonNull;

use crate::PriorityQueue;

type FibNodePtr<T> = NonNull<FibHeapNode<T>>;

pub struct FibHeapNode<T> {
    priority: usize,
    degree: usize,
    marked: bool,

    parent: Option<FibNodePtr<T>>,
    child: Option<FibNodePtr<T>>,
    left: FibNodePtr<T>,
    right: FibNodePtr<T>,

    data: T,
}

pub struct FibHeapPQ<T> {
    min: Option<FibNodePtr<T>>,
    size: usize,
}

impl<T> FibHeapPQ<T> {
    fn merge_with_root_list(&mut self, node: FibNodePtr<T>) {
        unsafe {
            if let Some(min) = self.min {
                let min_left = (*min.as_ptr()).left;

                (*node.as_ptr()).right = min;
                (*node.as_ptr()).left = min_left;
                (*min_left.as_ptr()).right = node;
                (*min.as_ptr()).left = node;
            } else {
                self.min = Some(node);
            }

            let is_new_min = self
                .min
                .is_none_or(|min| (*node.as_ptr()).priority < (*min.as_ptr()).priority);

            if is_new_min {
                self.min = Some(node);
            }
        }
    }

    fn remove_from_child_list(&mut self, parent: FibNodePtr<T>, node: FibNodePtr<T>) {
        unsafe {
            if (*parent.as_ptr()).child == Some(node) {
                (*parent.as_ptr()).child = if (*node.as_ptr()).right == node {
                    None
                } else {
                    Some((*node.as_ptr()).right)
                };
            }
            (*(*node.as_ptr()).left.as_ptr()).right = (*node.as_ptr()).right;
            (*(*node.as_ptr()).right.as_ptr()).left = (*node.as_ptr()).left;
            (*node.as_ptr()).left = node;
            (*node.as_ptr()).right = node;
            (*parent.as_ptr()).degree -= 1;
        }
    }

    fn remove_from_root_list(&mut self, node: FibNodePtr<T>) {
        unsafe {
            let left = (*node.as_ptr()).left;
            let right = (*node.as_ptr()).right;

            if right == node {
                self.min = None;
            } else {
                if self.min == Some(node) {
                    self.min = Some(right);
                }

                (*left.as_ptr()).right = right;
                (*right.as_ptr()).left = left;
            }

            (*node.as_ptr()).left = node;
            (*node.as_ptr()).right = node;
        }
    }

    fn cut(&mut self, l_node: FibNodePtr<T>, r_node: FibNodePtr<T>) {
        self.remove_from_child_list(r_node, l_node);
        self.merge_with_root_list(l_node);
        unsafe {
            (*l_node.as_ptr()).parent = None;
            (*l_node.as_ptr()).marked = false;
        }
    }

    fn cascading_cut(&mut self, node: FibNodePtr<T>) {
        unsafe {
            let ref_node = (*node.as_ptr()).parent;

            let Some(ref_node) = ref_node else {
                return;
            };
            if !(*node.as_ptr()).marked {
                (*node.as_ptr()).marked = true;
            } else {
                self.cut(node, ref_node);
                self.cascading_cut(ref_node);
            }
        }
    }

    fn heap_link(&mut self, l_node: FibNodePtr<T>, r_node: FibNodePtr<T>) {
        self.remove_from_root_list(l_node);
        unsafe {
            (*l_node.as_ptr()).left = l_node;
            (*l_node.as_ptr()).right = l_node;
            (*l_node.as_ptr()).parent = Some(r_node);
            (*l_node.as_ptr()).marked = false;

            if let Some(r_node_child) = (*r_node.as_ptr()).child {
                (*l_node.as_ptr()).left = r_node_child;
                (*l_node.as_ptr()).right = (*r_node_child.as_ptr()).left;
                (*(*r_node_child.as_ptr()).left.as_ptr()).right = l_node;
                (*r_node_child.as_ptr()).left = l_node;
            } else {
                (*r_node.as_ptr()).child = Some(l_node);
            }
            (*r_node.as_ptr()).degree += 1;
        }
    }

    fn consolidate(&mut self) {
        let Some(min) = self.min else {
            return;
        };
        if unsafe { (*min.as_ptr()).right == min } {
            return;
        }

        let max_degree = (self.size as f32).log2() as usize * 2 + 1;
        let mut degree_table: Vec<Option<FibNodePtr<T>>> = vec![None; max_degree];

        let mut roots = Vec::new();
        let mut cur = min;
        unsafe {
            loop {
                roots.push(cur);
                cur = (*cur.as_ptr()).right;
                if cur == min {
                    break;
                }
            }

            for x_ptr in roots {
                let mut x = x_ptr;
                let mut d = (*x.as_ptr()).degree;

                while let Some(y) = degree_table[d] {
                    let mut y = y;
                    if (*x.as_ptr()).priority > (*y.as_ptr()).priority {
                        std::mem::swap(&mut x, &mut y);
                    }
                    self.heap_link(y, x);
                    degree_table[d] = None;
                    d += 1;
                }
                degree_table[d] = Some(x);
            }

            self.min = None;
            for slot in degree_table.iter().flatten() {
                let node = slot.as_ptr();
                (*node).left = *slot;
                (*node).right = *slot;

                if let Some(cur_min) = self.min {
                    (*node).right = cur_min;
                    (*node).left = (*cur_min.as_ptr()).left;
                    (*(*cur_min.as_ptr()).left.as_ptr()).right = *slot;
                    (*cur_min.as_ptr()).left = *slot;

                    if (*node).priority < (*cur_min.as_ptr()).priority {
                        self.min = Some(*slot);
                    }
                } else {
                    self.min = Some(*slot);
                }
            }
        }
    }

    fn free_node_list(start: Option<FibNodePtr<T>>) {
        let Some(start_ptr) = start else {
            return;
        };

        let mut cur = start_ptr;
        unsafe {
            loop {
                let next = (*cur.as_ptr()).right;
                let child = (*cur.as_ptr()).child;
                Self::free_node_list(child);
                drop(Box::from_raw(cur.as_ptr()));
                if next == start_ptr {
                    break;
                }
                cur = next;
            }
        }
    }
}

impl<T> PriorityQueue<T> for FibHeapPQ<T> {
    type NodeIdentifier = FibNodePtr<T>;

    fn new() -> Self {
        Self { min: None, size: 0 }
    }

    // NOTE: with_capacity does not make sense for a fibonacci heap
    fn with_capacity(_capacity: usize) -> Self {
        Self::new()
    }

    fn insert(&mut self, data: T, priority: usize) -> Self::NodeIdentifier {
        let boxed = Box::new(FibHeapNode {
            priority,
            degree: 0,
            marked: false,
            parent: None,
            child: None,

            left: NonNull::dangling(),
            right: NonNull::dangling(),

            data,
        });

        let ptr = NonNull::new(Box::into_raw(boxed)).expect("Box::into_raw never returns null");

        unsafe {
            (*ptr.as_ptr()).left = ptr;
            (*ptr.as_ptr()).right = ptr;
        }

        self.merge_with_root_list(ptr);
        self.size += 1;

        ptr
    }

    fn extract_min(&mut self) -> Option<T> {
        let min = self.min?;

        unsafe {
            if let Some(start_child) = (*min.as_ptr()).child {
                let mut child = start_child;

                loop {
                    let next = (*child.as_ptr()).right;

                    (*child.as_ptr()).left = child;
                    (*child.as_ptr()).right = child;
                    (*child.as_ptr()).parent = None;
                    (*child.as_ptr()).marked = false;

                    self.merge_with_root_list(child);

                    child = next;

                    if child == start_child {
                        break;
                    }
                }

                (*min.as_ptr()).child = None;
                (*min.as_ptr()).degree = 0;
            }

            let next_root = (*min.as_ptr()).right;
            let was_singleton = next_root == min;

            self.remove_from_root_list(min);

            if was_singleton {
                self.min = None;
            } else {
                self.min = Some(next_root);
                self.consolidate();
            }

            self.size -= 1;

            let boxed = Box::from_raw(min.as_ptr());

            Some(boxed.data)
        }
    }

    #[inline]
    fn is_empty(&self) -> bool {
        self.size == 0
    }

    fn remove(&mut self, node: Self::NodeIdentifier) -> Option<T> {
        unsafe {
            (*node.as_ptr()).priority = 0;

            let parent = (*node.as_ptr()).parent;
            if let Some(parent) = parent {
                self.cut(node, parent);
                self.cascading_cut(parent);
            }
            self.min = Some(node);
        }

        self.extract_min()
    }

    fn decrease_key(&mut self, node: Self::NodeIdentifier, new_priority: usize) {
        unsafe {
            let node_ptr = node.as_ptr();

            if new_priority > (*node_ptr).priority {
                return;
            }

            (*node_ptr).priority = new_priority;

            if let Some(parent) = (*node_ptr).parent {
                let parent_ptr = parent.as_ptr();

                if (*node_ptr).priority < (*parent_ptr).priority {
                    self.cut(node, parent);
                    self.cascading_cut(parent);
                }
            }

            match self.min {
                Some(min) => {
                    if (*node_ptr).priority < (*min.as_ptr()).priority {
                        self.min = Some(node);
                    }
                }
                None => {
                    self.min = Some(node);
                }
            }
        }
    }

    fn peek(&self) -> Option<&T> {
        self.min
            .as_ref()
            .map(|node| unsafe { &(*node.as_ptr()).data })
    }

    fn merge(self, other: Self) -> Self {
        let mut new_pq = FibHeapPQ::<T> {
            min: self.min,
            size: self.size + other.size,
        };

        let (Some(lhs_min), Some(rhs_min)) = (self.min, other.min) else {
            new_pq.min = if self.min.is_some() {
                other.min
            } else {
                self.min
            };
            std::mem::forget(self);
            std::mem::forget(other);
            return new_pq;
        };

        unsafe {
            let lhs_last = (*lhs_min.as_ptr()).left;
            let rhs_last = (*rhs_min.as_ptr()).left;

            (*lhs_last.as_ptr()).right = rhs_min;
            (*rhs_min.as_ptr()).left = lhs_last;
            (*rhs_last.as_ptr()).right = lhs_min;
            (*lhs_min.as_ptr()).left = rhs_last;

            if (*rhs_min.as_ptr()).priority < (*lhs_min.as_ptr()).priority {
                new_pq.min = Some(rhs_min);
            }
        }

        std::mem::forget(self);
        std::mem::forget(other);
        new_pq
    }
}

impl<T> Drop for FibHeapPQ<T> {
    fn drop(&mut self) {
        let Some(start) = self.min else {
            return;
        };
        let mut cur = start;
        unsafe {
            loop {
                let next = (*cur.as_ptr()).right;
                FibHeapPQ::<T>::free_node_list((*cur.as_ptr()).child);
                drop(Box::from_raw(cur.as_ptr()));
                cur = next;
                if cur == start {
                    break;
                }
            }
        }
    }
}
